File: warmup/2015-europe-regional/testdata/elementarymath/submissions/accepted/tobi.java
import java.util.*;
import java.io.*;

public class tobi {
	public static ArrayList<ArrayList<Integer>> adj = new ArrayList<>();
	public static HashMap<Long, Integer> notoint = new HashMap<>();
	public static int N, nodes, pred[] = new int[7505];
	public static long a[] = new long[2505];
       	public static long b[] = new long[2505];
	public static long idtores[] = new long[7505];
	public static long res[] = new long[2505];
	public static boolean[] mark = new boolean[7505];

	public static boolean match(int n) {
		if (n == -1) return true;
		for (int to : adj.get(n)) {
			if (mark[to]) continue;
			mark[to] = true;
			if (match(pred[to])) {
				pred[to] = n;
				return true;
			}
		}
		return false;
	}

	public static int maxmatch() {
		Arrays.fill(pred, -1);
		int ret = 0;
		for (int i = 0; i < N; i++) {
			Arrays.fill(mark, false);
			if (match(i)) ret++;
		}
		return ret;
	}

	public static int get(long x) {
		if (notoint.get(x) == null) {
			notoint.put(x, nodes++);
			idtores[notoint.get(x)] = x;
		}
		return notoint.get(x);
	}

	public static void main(String args[]) {
		Scanner sc = new Scanner(System.in);
		N = sc.nextInt();
		for (int i = 0; i < N; i++) adj.add(new ArrayList<Integer>());
		for (int i = 0; i < N; i++) {
			a[i] = sc.nextInt();
			b[i] = sc.nextInt();
			adj.get(i).add(get(a[i]+b[i]));
			adj.get(i).add(get(a[i]*b[i]));
			adj.get(i).add(get(a[i]-b[i]));
		}

		if (maxmatch() < N) System.out.println("impossible");
		else {
			for (int i = 0; i < nodes; i++) if (pred[i] >= 0) res[pred[i]] = idtores[i];
			for (int i = 0; i < N; i++) {
				char c = '+';
				if (a[i]*b[i] == res[i]) c = '*';
				else if (a[i]-b[i] == res[i]) c = '-';
				System.out.println(a[i] + "" + c  + "" + b[i] + "=" + res[i]);
			}
		}
	}
}

File: warmup/2015-europe-regional/testdata/elementarymath/submissions/accepted/tobi.cpp
#include <bits/stdc++.h>

using namespace std;
typedef long long int ll;
#define FOR(i,a,b) for (ll i=(a); i<(b); i++)
vector<int> adj[2505];
map<ll,int> notoint;
int N, nodes, pred[7505];
ll a[2505], b[2505], idtores[7505], res[2505];
bool mark[7505];

bool match(int n) {
        if (n == -1) return true;
        for (int to : adj[n]) {
                if (mark[to]) continue;
                mark[to] = true;
                if (match(pred[to])) {
                        pred[to] = n;
                        return true;
                }
        }
        return false;
}

int maxmatch() {
        fill_n(pred, 7505, -1);
        int ret = 0;
        FOR(i, 0, N) {
                fill_n(mark, 7505, false);
                if (match(i)) ret++;
        }
        return ret;
}

int get(ll x) {
	if (!notoint.count(x)) {
		notoint[x] = nodes++;
		idtores[notoint[x]] = x;
	}
	return notoint[x];
}

int main() {
	cin >> N;
	FOR(i, 0, N) {
		cin >> a[i] >> b[i];
		adj[i].push_back(get(a[i]+b[i]));
		adj[i].push_back(get(a[i]*b[i]));
		adj[i].push_back(get(a[i]-b[i]));
	}

	if (maxmatch() < N) cout << "impossible" << endl;
	else {
		FOR(i, 0, nodes) if (pred[i] >= 0) res[pred[i]] = idtores[i];
		FOR(i, 0, N) {
			char c = '+';
			if (a[i]*b[i] == res[i]) c = '*';
			else if (a[i]-b[i] == res[i]) c = '-';
			cout << a[i] << c << b[i] << "=" << res[i] << endl;
		}
	}

	return 0;
}
